common/models/templates/src/priority-queue.ts

```diff
@@ -4,216 +4,215 @@
-    }
+  private static parentIndex(index: number): number {
+    return Math.floor((index-1)/2);
+  }
 
-    /**
-     * Maintains internal state, rearranging the internal state until all heap constraints
-     * are properly satisfied.
-     * - O(N) when 'heapifying' the whole heap
-     * - O(N) worst-case for partial heap operations (as part of an enqueueAll)
-     * <p>
-     */
-    private heapify(): void;
-    private heapify(start: number, end: number): void;
-    private heapify(start?: number, end?: number): void {
-      if(start == undefined || end == undefined) {
-        this.heapify(0, this.count - 1);
-      }
+  /**
+   * Maintains internal state, rearranging the internal state until all heap constraints
+   * are properly satisfied.
+   * - O(N) when 'heapifying' the whole heap
+   * - O(N) worst-case for partial heap operations (as part of an enqueueAll)
+   * <p>
+   */
+  private heapify(): void;
+  private heapify(start: number, end: number): void;
+  private heapify(start?: number, end?: number): void {
+    if(start == undefined || end == undefined) {
+      this.heapify(0, this.count - 1);
+      return;
+    }
 
-      // Use of 'indices' here is a bit of a customization.
-      // At the cost of (temporary) extra storage space, we can more efficiently enqueue
-      // multiple elements simultaneously.
-      let queuedIndices: number[] = [];
-      let lastParent = -1;
-
-      for(let i = end; i >= start; i--) {
-        let parent = PriorityQueue.parentIndex(i);
-        if(this.siftDown(i) && parent < start && lastParent != parent) {
-          // We only need to queue examination for a heap node if its children have changed
-          // and it isn't already being examined.
-          queuedIndices.push(parent);
-          lastParent = parent;
-        }
+    // Use of 'indices' here is a bit of a customization.
+    // At the cost of (temporary) extra storage space, we can more efficiently enqueue
+    // multiple elements simultaneously.
+    let queuedIndices: number[] = [];
+    let lastParent = -1;
+
+    for(let i = end; i >= start; i--) {
+      let parent = PriorityQueue.parentIndex(i);
+      if(this.siftDown(i) && parent < start && lastParent != parent) {
+        // We only need to queue examination for a heap node if its children have changed
+        // and it isn't already being examined.
+        queuedIndices.push(parent);
+        lastParent = parent;
       }
+    }
 
-      lastParent = -1;
-      while(queuedIndices.length > 0) {
-        let index = queuedIndices.shift();
-        let parent = PriorityQueue.parentIndex(index);
-        if(this.siftDown(index) && parent >= 0 && lastParent != parent) {
-            // We only need to queue examination for a heap node if its children have changed.
-          queuedIndices.push(parent);
-          lastParent = parent;
-        }
+    lastParent = -1;
+    while(queuedIndices.length > 0) {
+      let index = queuedIndices.shift() as number;
```

Comment:
Why is this change required?

Reply:
![image](https://user-images.githubusercontent.com/25213402/216245586-64d3cf9c-a607-4bb2-a115-eda39eb62bce.png)

TypeScript is failing to infer that the `| undefined` part is logically impossible here, which causes the following lines that expect a `number` to throw related errors.